### PR TITLE
API: Fix tags update on updating a note

### LINKED ISSUE #941: API : Tags not updated when updating a note
note creation 
```
curl -H 'Content-Type: application/json' --data '{"title":"my note", "body":"some markdown", "tags": "tags1,tag2,tag3"}' http://127.0.0.1:41184/notes
{"title":"my note","body":"some markdown","parent_id":"0882e5b783b94f61baa428273147fbbf","source":"joplin-desktop","source_application":"net.cozic.joplin-desktop","id":"a315850d76b54bd5960e05042f976eb7"
```
updating the tags looks fine 
```
 curl -X PUT -H 'Content-Type: application/json' --data '{"title":"my note", "body":"some markdown", "tags": "tags1,tag2,tag3, tag4"}' http://127.0.0.1:41184/notes/a315850d76b54bd5960e05042f976eb7/?token=my_long_token
{"id":"a315850d76b54bd5960e05042f976eb7","parent_id":"0882e5b783b94f61baa428273147fbbf","title":"my note","body":"some markdown","created_time":1541081002837,"updated_time":1541081146841,"is_conflict":0,"latitude":"0.00000000","longitude":"0.00000000","altitude":"0.0000","author":"","source_url":"","is_todo":0,"todo_due":0,"todo_completed":0,"source":"joplin-desktop","source_application":"net.cozic.joplin-desktop","application_data":"","order":0,"user_created_time":1541081002837,"user_updated_time":1541081146841,"encryption_cipher_text":"","encryption_applied":0,"type_":1,"tags":"tags1,tag2,tag3, tag4"}
```
but if I look at the database 
```
sqlite> select tags.title from note_tags,tags where note_id = 'a315850d76b54bd5960e05042f976eb7' and tag_id = tags.id;
tags1
tag2
tag3
```
no `tag4`

### PR DESCRIPTION
**Pull request by jyuvaraj03**
Fixes #941 

**Before:**
before-2020-03-04-002019

**After:**
after-2020-03-04-000934

The issue as mentioned in #941 was that the tags were not updated. This was due to the fact that in `PUT` request method the tags were not explicitly set unlike how the `POST` request method was handled in the `action_notes` function. 

The fix I have given makes the necessary changes. I have also written unit tests for the code. Looking forward to any comments or suggestions!

### CODE REVIEW

**Comment by jyuvaraj03** [comment]
@PackElend label me please

File: ReactNativeClient/lib/services/rest/Api.js [hunk 1]
```diff
@@ -451,6 +451,23 @@ class Api {
 			return note;
 		}
 
+		if (request.method === 'PUT') {
+			const note = await Note.load(id);
+			const requestBody = JSON.parse(request.body);
+
+			if (!note) throw new ErrorNotFound();
+
+			let newNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: I'm not familiar with this interface so please help me understand. Is the `PUT` command intended to update an existing note?  Is this really a new note, or an updated one?

File: ReactNativeClient/lib/services/rest/Api.js [hunk 2]
```diff
@@ -451,6 +451,23 @@ class Api {
...
+			newNote = await Note.save(newNote, { userSideValidation: true });
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: What is the effect of user-side validation here?

File: CliClient/tests/services_rest_Api.js [hunk 3]
```diff
@@ -326,4 +326,28 @@ describe('services_rest_Api', function() {
 		expect(response3.length).toBe(2);
 	}));
 
+	it('should update tags when updating notes', asyncTest(async () => {
+		const tag1 = await Tag.save({ title: 'mon étiquette 1' });
+		const tag2 = await Tag.save({ title: 'mon étiquette 2' });
+		const tag3 = await Tag.save({ title: 'mon étiquette 3' });
+		const note = await Note.save({
+			title: 'ma note un',
+			tags: `${tag1.title},${tag2.title}`,
```
Comment (mic704b) [inline CliClient/tests/services_rest_Api.js]: Does this add the tags correclty?  Perhaps you could start the testing by getting the note tags to check the note is set up as you intend?

File: CliClient/tests/services_rest_Api.js [hunk 4]
```diff
@@ -326,4 +326,28 @@ describe('services_rest_Api', function() {
...
+			tags: `${tag1.title}`,
+		}));
+		expect(response2.tags === `${tag1.title}`).toBe(true);
+
+		const response3 = await api.route('PUT', `notes/${note.id}`, null, JSON.stringify({
+			tags: `${tag1.title},${tag2.title},${tag3.title}`,
+		}));
+		expect(response3.tags === `${tag1.title},${tag2.title},${tag3.title}`).toBe(true);
+	}));
 });
```
Comment (mic704b) [inline CliClient/tests/services_rest_Api.js]: What happens if we accidently add a tag name that doesn't exist? (Is it possible to do via this interface?  If it is, then perhaps a test would be good.)

File: ReactNativeClient/lib/services/rest/Api.js (hunk 1 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/services/rest/Api.js]: Yes, the `PUT` command is intended to update an existing note. Perhaps, the variable name `newNote` was a bit confusing? I will change it to `updatedNote`.

File: CliClient/tests/services_rest_Api.js (hunk 3 above)
Comment (jyuvaraj03) [reply CliClient/tests/services_rest_Api.js]: Yes, it does add the tags in the note (checked in the desktop app). But I haven't checked the note in the unit test. I will rewrite the unit test to make sure it tests the notes rather than the response.

File: CliClient/tests/services_rest_Api.js (hunk 4 above)
Comment (jyuvaraj03) [reply CliClient/tests/services_rest_Api.js]: If the user sets a tag name that doesn't exist via this interface, it creates the tags and then adds it to the note. I forgot to write a test case for that. I will fix the unit test.

File: ReactNativeClient/lib/services/rest/Api.js (hunk 2 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/services/rest/Api.js]: The code I have written here was actually refactored from the `defaultAction_` function which was previously called in the case of `PUT` methods. 

To be honest, I do not know exactly know what user-side validation does here. But since it was already present in the code that was executed (before this fix), I decided to leave it untampered in order not to break any functionality.

**Comment by jyuvaraj03** [comment]
> Should you also check the actual tags on the note note after each `PUT` (using `Tag` and/or `Note` methods to directly interrogate the associated tags)? I'm not sure the test is covering the issue.

Yes, you were right. I should have checked the actual tags on the notes themselves. I have since rewritten the unit tests to do this.

File: CliClient/tests/services_rest_Api.js (hunk 3 above)
Comment (mic704b) [reply CliClient/tests/services_rest_Api.js]: Since you have the response checks, they are still useful.  So just add the new checks, so it checks both.

**Comment by jyuvaraj03** [comment]
I have done all the requested changes. Please review :)

File: ReactNativeClient/lib/services/rest/Api.js [hunk 5]
```diff
@@ -451,6 +451,23 @@ class Api {
 			return note;
 		}
 
+		if (request.method === 'PUT') {
+			const note = await Note.load(id);
+			const requestBody = JSON.parse(request.body);
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: For consistency we should change the name of 'requestBody' to match what is used in the 'POST' code (just above).

Also there is no need to create this so early in the function, so please move it closer to where it is first used.

File: ReactNativeClient/lib/services/rest/Api.js [hunk 6]
```diff
@@ -451,6 +451,23 @@ class Api {
...
+
+			if (!note) throw new ErrorNotFound();
+
+			let updatedNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
+			updatedNote = await Note.save(updatedNote, { userSideValidation: true });
+
+			if (requestBody.tags) {
+				const tagTitles = requestBody.tags.split(',');
+				await Tag.setNoteTagsByTitles(updatedNote.id, tagTitles);
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: `updatedNote.id` can be just `id`

File: ReactNativeClient/lib/services/rest/Api.js [hunk 7]
```diff
@@ -451,6 +451,23 @@ class Api {
...
 		}
 
+		if (request.method === 'PUT') {
+			const note = await Note.load(id);
+			const requestBody = JSON.parse(request.body);
+
+			if (!note) throw new ErrorNotFound();
+
+			let updatedNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
+			updatedNote = await Note.save(updatedNote, { userSideValidation: true });
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: You mentioned you borrowed this from the default handler.  So rather than duplicate the code here, why not call the default handler to do this?

Comment (jyuvaraj03) [reply ReactNativeClient/lib/services/rest/Api.js]: Yeah, I didn't think about that! I'll implement it now.

File: ReactNativeClient/lib/services/rest/Api.js (hunk 5 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/services/rest/Api.js]: > Also there is no need to create this so early in the function, so please move it closer to where it is first used.

I have made this change. But, can I know why it is important to not create it early in the function?

**Comment by jyuvaraj03** [comment]
Completed all the requested changes. Please review!

**Comment by laurent22** [comment]
@mic704b, I've sent you an invite to be a collaborator on the project, so when you think this one is ready feel free to merge it.

Please use the syntax `Platform: Fixes|Resolves #xxx: Comment`. The comment will appear in the changelog so it should be something not too technical (indeed the title of this PR would be good). So for example `Api: Resolves #941: Fix tags update on updating a note`

**Comment by jyuvaraj03** [comment]
Sorry, I accidentally requested a review! Please review at your own pace.

**Comment by mic704b** [comment]
> @mic704b, I've sent you an invite to be a collaborator on the project, so when you think this one is ready feel free to merge it.

Thank you @laurent22! I am honoured to be invited.

File: CliClient/tests/services_rest_Api.js [hunk 8]
```diff
@@ -326,4 +327,33 @@ describe('services_rest_Api', function() {
...
+
+		const response2 = await api.route('PUT', `notes/${note.id}`, null, JSON.stringify({
+			tags: `${tag2.title},${newTagTitle}`,
+		}));
+		const newTag = await Tag.loadByTitle(newTagTitle);
+		const tagIds2 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response2.tags === `${tag2.title},${newTag.title}`).toBe(true);
+		expect(tagIds2.includes(tag2.id)).toBe(true);
+		expect(tagIds2.includes(newTag.id)).toBe(true);
+	}));
```
Comment (mic704b) [inline CliClient/tests/services_rest_Api.js]: Please add a test case for when tags are an empty string.  It is not expected, but we could receive it, so we should test we handle it properly.

File: CliClient/tests/services_rest_Api.js [hunk 9]
```diff
@@ -326,4 +327,33 @@ describe('services_rest_Api', function() {
...
+		Tag.addNote(tag1.id, note.id);
+		Tag.addNote(tag2.id, note.id);
+
+		const response1 = await api.route('PUT', `notes/${note.id}`, null, JSON.stringify({
+			tags: `${tag1.title},${tag3.title}`,
+		}));
+		const tagIds1 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response1.tags === `${tag1.title},${tag3.title}`).toBe(true);
+		expect(tagIds1.includes(tag1.id)).toBe(true);
+		expect(tagIds1.includes(tag3.id)).toBe(true);
```
Comment (mic704b) [inline CliClient/tests/services_rest_Api.js]: We should also confirm what happened to tag2. Its absence can be confirmed by adding a check on the length of `tagIds1`

**Comment by mic704b** [comment]
@jyuvaraj03 looks good, nearly there.

**Comment by jyuvaraj03** [comment]
> Please add a test case for when tags are an empty string. It is not expected, but we could receive it, so we should test we handle it properly.

Thanks for pointing it out. That was an excellent catch! I hadn't handled that case. When the tags in the request is empty, the tags associated with the note do not get deleted like intended. Instead, they stay the same. 

So I handled it by checking if the `tags` property is explicitly given as `''` during the API request. If so, all the existing tags are removed from the corresponding note. If the `tags` property is not mentioned during the request, no change is made to the tags in the note.

File: CliClient/tests/services_rest_Api.js [hunk 10]
```diff
@@ -326,4 +327,50 @@ describe('services_rest_Api', function() {
...
+		expect(tagIds3.includes(tag2.id)).toBe(true);
+		expect(tagIds3.includes(newTag.id)).toBe(true);
+
+		const response4 = await api.route('PUT', `notes/${note.id}`, null, JSON.stringify({
+			tags: '',
+		}));
+		const tagIds4 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response4.tags === '').toBe(true);
+		expect(tagIds4.length === 0).toBe(true);
+	}));
```
Comment (jyuvaraj03) [inline CliClient/tests/services_rest_Api.js]: In this test case, the `tags` property is mentioned explicitly with a `''`. So, all the tags associated with the note are removed from the note.

File: CliClient/tests/services_rest_Api.js [hunk 11]
```diff
@@ -326,4 +327,50 @@ describe('services_rest_Api', function() {
...
+		expect(tagIds2.includes(tag2.id)).toBe(true);
+		expect(tagIds2.includes(newTag.id)).toBe(true);
+
+		const response3 = await api.route('PUT', `notes/${note.id}`, null, JSON.stringify({
+			title: 'Some other title',
+		}));
+		const tagIds3 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(tagIds3.length === 2).toBe(true);
+		expect(tagIds3.includes(tag2.id)).toBe(true);
+		expect(tagIds3.includes(newTag.id)).toBe(true);
```
Comment (jyuvaraj03) [inline CliClient/tests/services_rest_Api.js]: This test case checks the case when the `tags` property is **not** mentioned during the request. So, the tags associated with the note does not differ from the previous test case.

Also, I haven't checked `response3.tags` because it is `undefined` when `tags` property isn't mentioned during the API request. So, I skipped that check.

File: ReactNativeClient/lib/services/rest/Api.js [hunk 12]
```diff
@@ -451,6 +451,24 @@ class Api {
...
+
+			let updatedNote = await this.defaultAction_(BaseModel.TYPE_NOTE, request, id, link);
+
+			const requestNote = JSON.parse(request.body);
+			if (requestNote.tags === '') {
+				await Tag.setNoteTagsByTitles(id, '');
+			} else if (requestNote.tags) {
+				const tagTitles = requestNote.tags.split(',');
+				await Tag.setNoteTagsByTitles(id, tagTitles);
+			}
```
Comment (mic704b) [inline ReactNativeClient/lib/services/rest/Api.js]: Can this be reduced to a single if statement?  I don't think you need two branches.

File: CliClient/tests/services_rest_Api.js (hunk 10 above)
Comment (mic704b) [inline CliClient/tests/services_rest_Api.js]: Generally, for future maintenance, it would be better to separate the test cases into 3 or 4 smaller test cases rather than one big one. This also helps avoid using the 1,2,3,4 variable names too.  Sometimes it is too inefficient, but by default it is preferred.

File: CliClient/tests/services_rest_Api.js (hunk 11 above)
Comment (mic704b) [reply CliClient/tests/services_rest_Api.js]: > because it is `undefined` when `tags` property isn't mentioned

That can be tested too!  If it wasn't returned as undefined, then it would be a problem.

**Comment by laurent22** [comment]
> One more question: does the API documentation need update? The documentation says PUT on a note is to set note properties, but doesn't list tags as a property. Perhaps it should be added to the list, or the description changed to something like `Sets the properties and tags of ...`. What do you think?

That's a good point, the doc should be updated. To do so @jyuvaraj03 you can update command-apidoc.js, then you can verify the output from /CliClient using `npm start -- apidoc`.

In fact, ideally we should support all the special properties that the POST method supports, such as body_html, image_data_url, crop_rect, etc. for consistency. That could be done with a bit of refactoring, but it's not needed for this PR.

File: ReactNativeClient/lib/services/rest/Api.js (hunk 12 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/services/rest/Api.js]: Yes, I just realized two branches weren't necessary. I changed the code to use just a single `if`.

File: CliClient/tests/services_rest_Api.js (hunk 10 above)
Comment (jyuvaraj03) [reply CliClient/tests/services_rest_Api.js]: I restructured all the test cases into separate ones. But, I'm not sure if I did it right. I have used lengthy test case descriptions for some test cases. Please review and let me know if I'm doing it right.

**Comment by jyuvaraj03** [comment]
> One more question: does the API documentation need update? The documentation says PUT on a note is to set note properties, but doesn't list tags as a property. Perhaps it should be added to the list, or the description changed to something like `Sets the properties and tags of ...`. What do you think?

You're right. So, I updated the doc. I added the `tags` property to the list of properties.

**Comment by jyuvaraj03** [comment]
> That's a good point, the doc should be updated. To do so @jyuvaraj03 you can update command-apidoc.js, then you can verify the output from /CliClient using npm start -- apidoc.

Yes, I did update the command-apidoc.js file. But, I have a small question. These were the lines of code I changed in that file:
```
tableFields.push({
	name: 'tags',
	type: Database.enumId('fieldType', 'text'),
	description: 'Comma-separated list of tags. eg. `tag1,tag2`.',
});
```

I just uncommented them in this commit. These lines already existed, but you had commented them out. Was there a reason you did that? Am I missing something?

<hr> 

> In fact, ideally we should support all the special properties that the POST method supports, such as body_html, image_data_url, crop_rect, etc. for consistency. That could be done with a bit of refactoring, but it's not needed for this PR.

I would like to work on implementing these (possibly in another PR). How should I go about it? Should I create an issue for this and then submit a PR?

**Comment by laurent22** [comment]
Ok I think it's ready to merge now, thanks @jyuvaraj03 and thanks @mic704b for reviewing!

File: CliClient/tests/services_rest_Api.js (hunk 10 above)
Comment (mic704b) [reply CliClient/tests/services_rest_Api.js]: Looks good!

**Comment by mic704b** [comment]
> I think it's ready to merge now

Definitely.

Good work @jyuvaraj03
